### PR TITLE
Add support for floats

### LINKED ISSUE #59: Float literals
We currently support int literals, but we should also support float literals

### PR DESCRIPTION
**Pull request by gregorybchris**
- Add `FloatLit`
- Update `Lexer` to support tokenizing numbers with decimal parts
- Add `Float` object
- Update `Parser` to parse `FloatLit` into `Float`
- Update arithmetic binops to work for both `Int` and `Float`
- Add matching support for `Float` objects
- Add serialize/bencode support for `Float` objects

### CODE REVIEW

File: scrapscript.py [hunk 1]
```diff
@@ -37,10 +37,15 @@ class Token:
 
 
 @dataclass(eq=True)
-class NumLit(Token):
+class IntLit(Token):
```
Comment (gregorybchris) [inline scrapscript.py]: Can do this rename from `NumLit` to `IntLit` in another PR if we want

File: scrapscript.py [hunk 2]
```diff
@@ -206,11 +211,21 @@ def read_comment(self) -> None:
             pass
 
     def read_number(self, first_digit: str) -> Token:
+        # TODO: Support floating point numbers with no integer part
```
Comment (gregorybchris) [inline scrapscript.py:214]: There's also e notation/scientific notation to think about, but I think that can be a separate PR

File: scrapscript.py [hunk 3]
```diff
@@ -1005,19 +1038,25 @@ def make_bool(x: bool) -> Object:
     return Symbol("true" if x else "false")
 
 
+def wrap_inferred_number_type(x: Union[int, float]) -> Object:
```
Comment (gregorybchris) [inline scrapscript.py:1043]: My only concern with this is that it relies on the Python language quite a bit. If this is meant to be a reference implementation that can be ported to other languages I'm not sure if we want implement some of this arithmetic type inference from scratch.

File: scrapscript.py (hunk 1 above)
Comment (tekknolagi) [reply scrapscript.py]: I think in a separate commit, not necessarily separate PR

Comment (gregorybchris) [reply scrapscript.py]: Eh, I'll do a separate PR to keep things cleaner: #69

Comment (gregorybchris) [reply scrapscript.py]: Done. This PR diff should be easier to read through now

File: scrapscript.py (hunk 3 above)
Comment (tekknolagi) [reply scrapscript.py:1043]: This is a really good callout

I don't have a good idea for how to do this with few lines of code and I don't love the current code

File: scrapscript.py [hunk 4]
```diff
@@ -1190,6 +1231,8 @@ def bencode(obj: object) -> bytes:
     assert not isinstance(obj, bool)
     if isinstance(obj, int):
         return b"i" + str(int(obj)).encode("ascii") + b"e"
+    if isinstance(obj, float):
+        return b"f" + str(float(obj)).encode("ascii") + b"e"
```
Comment (tekknolagi) [inline scrapscript.py]: This is not standard bencode; probably we should serialize them as strings (they are tagged at a higher level)

Comment (tekknolagi) [reply scrapscript.py]: Or we should dump them as bytestrings or something since they're all 8 bytes wide

File: scrapscript.py [hunk 5]
```diff
@@ -1219,6 +1262,12 @@ def decode_int(self) -> int:
             buf += c
         return int(buf)
 
+    def decode_float(self) -> float:
+        buf = ""
+        while (c := self.read()) != "e":
+            buf += c
+        return float(buf)
+
```
Comment (tekknolagi) [inline scrapscript.py]:
```suggestion
```

File: scrapscript.py [hunk 6]
```diff
@@ -1251,6 +1300,8 @@ def decode(self) -> object:
         ty = self.read()
         if ty == "i":
             return self.decode_int()
+        if ty == "f":
+            return self.decode_float()
```
Comment (tekknolagi) [inline scrapscript.py]:
```suggestion
```

File: scrapscript.py [hunk 7]
```diff
@@ -1285,6 +1336,12 @@ def test_tokenize_multiple_digits(self) -> None:
     def test_tokenize_negative_int(self) -> None:
         self.assertEqual(tokenize("-123"), [Operator("-"), IntLit(123)])
 
+    def test_tokenize_float(self) -> None:
+        self.assertEqual(tokenize("3.14"), [FloatLit(3.14)])
+
+    def test_tokenize_negative_float(self) -> None:
+        self.assertEqual(tokenize("-3.14"), [Operator("-"), FloatLit(3.14)])
+
```
Comment (tekknolagi) [inline scrapscript.py:1338]: Add tests for `1.`, `.1`, `1.0.2`, `1..`, etc

File: scrapscript.py [hunk 8]
```diff
@@ -2122,6 +2185,15 @@ def test_match_with_inequal_ints_returns_none(self) -> None:
     def test_match_int_with_non_int_returns_none(self) -> None:
         self.assertEqual(match(String("abc"), pattern=Int(1)), None)
 
+    def test_match_with_equal_floats_returns_empty_dict(self) -> None:
```
Comment (tekknolagi) [inline scrapscript.py]: I don't know that we should support direct match on floats. Float equality can be very finicky. Perhaps we should have a way to do with some epsilon

File: scrapscript.py [hunk 9]
```diff
@@ -3547,6 +3641,9 @@ class BencodeTests(unittest.TestCase):
     def test_bencode_int(self) -> None:
         self.assertEqual(bencode(123), b"i123e")
 
+    def test_bencode_float(self) -> None:
+        self.assertEqual(bencode(3.14), b"f3.14e")
+
```
Comment (tekknolagi) [inline scrapscript.py]:
```suggestion
```

File: scrapscript.py [hunk 10]
```diff
@@ -3579,6 +3676,12 @@ def test_bdecode_int(self) -> None:
     def test_bdecode_negative_int(self) -> None:
         self.assertEqual(bdecode("i-123e"), -123)
 
+    def test_bdecode_float(self) -> None:
+        self.assertEqual(bdecode("f3.14e"), 3.14)
+
+    def test_bdecode_negative_float(self) -> None:
+        self.assertEqual(bdecode("f-3.14e"), -3.14)
+
```
Comment (tekknolagi) [inline scrapscript.py]:
```suggestion
```

**Review by tekknolagi — COMMENTED** [review]
I think there is more to be worked out for a clean implementation of the math operators

File: scrapscript.py (hunk 4 above)
Comment (tekknolagi) [reply scrapscript.py]: Although, hmm, there is apparently an extension: https://github.com/mgorny/gx86-test/blob/master/net-p2p/bittorrent/files/bittorrent-4.4.0-bencode-float.patch

File: scrapscript.py (hunk 8 above)
Comment (gregorybchris) [reply scrapscript.py]: Probably worth doing whatever other languages with pattern matching do if it's easy. I could imagine some languages might just allow you to do it and you should know that sometimes floats will be not quite equal and you won't get a match. If it's not easy I'll just remove pattern matching for floats here.

File: scrapscript.py (hunk 4 above)
Comment (gregorybchris) [reply scrapscript.py]: Oh cool. Looks like they replace the e for scientific notation with a capital E to avoid clashing with the bencode end sentinel e

Comment (gregorybchris) [reply scrapscript.py]: Do you like that? Could also just not support float serialization for now

Comment (tekknolagi) [reply scrapscript.py]: I am tentatively on team not-serialize-for-now but if you want to go ahead and do the full rigamarole for porting that impl and working on tests and stuff, go for it

File: scrapscript.py (hunk 3 above)
Comment (gregorybchris) [reply scrapscript.py:1043]: My best idea is to just add a to-do. even if we had one implementation for int and one for float for each binop that's quite a bit of code for some simple operators. And then if you want to be able to multiply/divide/add etc ints with floats it gets even more complex.

I think the alternative is just to not support floats until we have some fancier multiple dispatch..

File: scrapscript.py (hunk 4 above)
Comment (gregorybchris) [reply scrapscript.py]: We're on the same team. It's a shame to not be able to use floats in the online repl, but we can have a nice error message that says floats aren't supported rather than getting some cryptic syntax error like we get now when trying to use floats

File: scrapscript.py (hunk 7 above)
Comment (gregorybchris) [reply scrapscript.py:1338]: Good call, thanks

File: scrapscript.py (hunk 8 above)
Comment (gregorybchris) [reply scrapscript.py]: In Rust matching on floats is a warning that will (soon?) be turned into an error. 
https://github.com/rust-lang/rust/issues/41255
https://github.com/rust-lang/rust/issues/

In Haskell, Dart, and some others matching on floats seems supported.

I think it's fair to just not support pattern matching on floats for now since it's not so pressing

Comment (gregorybchris) [reply scrapscript.py]: Removed

**Review by tekknolagi — APPROVED** [review]
noice

**Comment by tekknolagi** [comment]
please "squash and merge" using github ui
